vm/src/builtins.rs

```diff
@@ -101,8 +101,16 @@ fn builtin_bin(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
 
 fn builtin_callable(vm: &mut VirtualMachine, args: PyFuncArgs) -> PyResult {
     arg_check!(vm, args, required = [(obj, None)]);
-    // TODO: is this a sufficiently thorough check?
-    let is_callable = obj.has_attr("__call__");
+    let is_callable = match obj.borrow().payload {
```

Comment:
I understand that we require to check here. But actually, the has_attr call should be sufficient. We might consider giving all types a table with methods. @cthulahoops what do you say?

Reply:
Yes, I think the has_attr call should be sufficient. Functions, etc should have a ```__call__```.

Longer term we should have a table for direct lookup of methods (like ```->tp_call``` in CPython).

Reply:
Not sure if this is considered a blocking issue?
I assumed this would work as it is, and get refactored when a common lookup table is added.

Reply:
Hi, sorry for being unclear.

I don't think this change is correct. The bug is that the various function and method types don't implement ```__call__```.

The following code fails when it shouldn't:

```
>>>>> (lambda: 7).__call__()
Traceback (most recent call last):
  File <stdin>, line 0, in <module>
AttributeError: 'function' object has no attribute '__call__'
```

If we add ```__call__``` to those types then this function won't need to be changed. (We can possibly simplify ```vm.invoke``` too, though I've got a feeling there's some annoying circular issues there.)

Reply:
Alright, so the intention is to use the target implementation, that while not working for all types now, will start working after `__call__` is fixed. I get it now.
> If we add `__call__` to those types then this function won't need to be changed.

AFAIK that's not entirely right. Based on this:
```
class C: pass
c = C()
c.__call__ = 1
assert not callable(c)
```
At the minimum, `callable()` will need to be changed from `obj.has_attr("__call__")` to `obj.typ().has_attr("__call__")`. Will do that fix soon.

Reply:
Yes, that's a good point.

Reply:
Updated.